src/edit_mode/vi/mod.rs

```diff
@@ -158,3 +162,96 @@ impl Vi {
         ReedlineEvent::Edit(output)
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use pretty_assertions::assert_eq;
+    use rstest::rstest;
+
+    fn edit_event(cmd: EditCommand) -> ReedlineEvent {
+        ReedlineEvent::Edit(vec![cmd])
+    }
+
+    fn char_key_event(ch: char) -> Event {
+        Event::Key(KeyEvent::new(KeyCode::Char(ch), KeyModifiers::NONE))
+    }
+
+    #[rstest]
+    #[case('h', edit_event(EditCommand::MoveLeft))]
+    #[case('j', ReedlineEvent::Down)]
+    #[case('k', ReedlineEvent::Up)]
+    #[case('l', edit_event(EditCommand::MoveRight))]
+    #[case('u', edit_event(EditCommand::Undo))]
+    #[case('p', edit_event(EditCommand::PasteCutBuffer))]
+    #[case('w', edit_event(EditCommand::MoveWordRight))]
+    #[case('b', edit_event(EditCommand::MoveWordLeft))]
+    #[case('0', edit_event(EditCommand::MoveToStart))]
+    #[case('$', edit_event(EditCommand::MoveToEnd))]
+    #[case('A', edit_event(EditCommand::MoveToEnd))] // Not checking if it also moves to end
+    #[case('D', edit_event(EditCommand::CutToEnd))]
```

Comment:
I really like how neatly and easily we can write tests with rstest. Let me know what others think.

Aside: One nice thing about these tests is that even if we switch to a different vi parsing logic (as outlined in parser.rs) this interface is not going to change and these tests will all remain valid and hopefully might help in dome debugging.